R/SEIaImIsRD.R

```diff
@@ -0,0 +1,356 @@
+#' omega - rate at which recovered individuals become susceptible,
+#' p_symptom - a list of fraction of different infected groups,
+#' gamma - a list of the rate of removal of each infected group (i.e. recovery rate of an infected individual),
+#' mu - a list of the rate of disease-caused mortality of each infected group
+#' @return An object of class SEIaImIsRD with initial population and parameters
+#' @aliases transmission_parameters<-,SEIaImIsRD-method
+#' @export
+#'
+setMethod(
+  "transmission_parameters<-",
+  "SEIaImIsRD",
+  function(
+    object,
+    value = list(beta = list(), kappa = NA_real_, omega = NA_real_,
+    p_symptom = list(), gamma = list(), mu = list())) {
+    param_list <- value
+    names(param_list) <- object@transmission_parameter_names
+    object@transmission_parameters <- param_list
+    # check if values are valid
+    errors <- character()
+    # check whether all required parameters are set
+    is_na_params <- is.na(param_list)
+    if (sum(is_na_params) != 0) {
+      msg <- paste("Missing parameters:",
+                   paste(names(param_list)[is_na_params], collapse = ", ")
+      )
+      errors <- c(errors, msg)
+    }
+    # check whether the lengths of beta, gamma and mu correspond to the number of infected groups
+    n_beta <- length(param_list$beta)
+    if (n_beta != 3) {
+      msg <- paste0(
+        "Length of parameter beta,", n_beta, ", is not equal to the setting ", 3)
+      errors <- c(errors, msg)
+    }
+    n_gamma <- length(param_list$gamma)
+    if (n_gamma != 3) {
+      msg <- paste0(
+        "Length of parameter gamma,", n_gamma, ", is not equal to the setting ", 3)
+      errors <- c(errors, msg)
+    }
+    n_mu <- length(param_list$mu)
+    if (n_mu != 3) {
+      msg <- paste0(
+        "Length of parameter mu,", n_mu, ", is not equal to the setting ", 3)
+      errors <- c(errors, msg)
+    }
+    if (length(errors) == 0) {
+      object@transmission_parameters <- param_list
+      object
+      }
+    else stop(paste(errors, ", please check and rerun transmission_parameters<-.\n"))
+    })
+
+#' Solves the ode system.
+#'
+#' \deqn{\frac{dS(t)}{dt} = -S(t)(\beta.i_{asymptomatic} I_{asymptomatic}(t) + \beta.i_{mild} I_{mild}(t) + \beta.i_{severe} I_{severe}(t)) + \omega R(t)}
+#' \deqn{\frac{dE(t)}{dt} = S(t)(\beta.i_{asymptomatic} I_{asymptomatic}(t) + \beta.i_{mild} I_{mild}(t) + \beta.i_{severe} I_{severe}(t)) - \kappa E(t)}
+#' \deqn{\frac{dI_{asymptomatic}(t)}{dt} = p_symptom.i_{asymptomatic} \kappa E(t) - (\gamma.i_{asymptomatic} + \mu.i_{asymptomatic}) I_{asymptomatic}(t)}
+#' \deqn{\frac{dI_{mild}(t)}{dt} = p_symptom.i_{mild}\kappa E(t) - (\gamma.i_{mild} + \mu.i_{mild}) I_{mild}(t)}
+#' \deqn{\frac{dI_{severe}(t)}{dt} = p_symptom.i_{severe}\kappa E(t) - (\gamma.i_{severe} + \mu.i_{severe}) I_{severe}(t)}
+#' \deqn{\frac{dR(t)}{dt} = -\omega R(t) + \gamma.i_{mild} I_{mild}(t) + \gamma.i_{severe} I_{severe}(t)}
```

Comment:
Shouldn't asymptomatics also recover?

Reply:
Think this is just a typo as it's in the ODE itself?

Reply:
Yes it is a typo in the documentations. I've fixed it. the actual functions were correct.